src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMultiMap.java

```diff
@@ -96,35 +93,74 @@ public void add(K k, V v, Handler<AsyncResult<Void>> completionHandler) {
 
   @Override
   public void get(K k, Handler<AsyncResult<ChoosableIterable<V>>> resultHandler) {
-    ChoosableSet<V> entries = cache.get(k);
-    if (entries != null && entries.isInitialised() && getInProgressCount.get() == 0) {
-      resultHandler.handle(Future.succeededFuture(entries));
-    } else {
-      getInProgressCount.incrementAndGet();
-      vertx.getOrCreateContext().<ChoosableIterable<V>>executeBlocking(fut -> {
-        Collection<V> entries2 = map.get(k);
-        ChoosableSet<V> sids;
-        if (entries2 != null) {
-          sids = new ChoosableSet<>(entries2.size());
-          for (V hid : entries2) {
-            sids.add(hid);
-          }
-        } else {
-          sids = new ChoosableSet<>(0);
+    ContextImpl context = vertx.getOrCreateContext();
+    @SuppressWarnings("unchecked")
+    Queue<GetRequest<K, V>> getRequests = (Queue<GetRequest<K, V>>) context.contextData().computeIfAbsent(this, ctx -> new ArrayDeque<>());
+    synchronized (getRequests) {
+      ChoosableSet<V> entries = cache.get(k);
+      if (entries != null && entries.isInitialised() && getRequests.isEmpty()) {
```

Comment:
can't the getrequests be avoided all together by marking the entries by some kind of a requestId should should fulfill its synchronization and only requests hitting keys with some in-progress operation should wait. Something like

if(entries != null && entries.pendingOperationId() != null)
   ...enqueue...
else
    return from cache.